### DISCUSSION TITLE
Is it possible to use existing Vulkan context?

### DISCUSSION THREAD

**Discussion by trsh** [LuisaGroup/LuisaCompute#125] [Q&A] [answered]
I know, vulkan is still in works, but would this be possible?

**Comment by Mike-Leo-Smith** [comment]
Yes. I can make the `VulkanSwapchain` interface class public so you can query the underlying `VkInstance`/`VkPhysicalDevice`/etc.

**Reply by trsh** [reply]
I think there is a misunderstanding. What I mean was, that I already have phy_device, device and vk_instance from main app, and I want to Initialize LuisaCompute with it.

**Reply by Mike-Leo-Smith** [reply]
Ah, that would need some tricks to make the LC backends aware of the existing Vulkan objects. Maybe I can provide some options in `DeviceConfig` so that you can pass the objects when creating a Luisa device?

**Reply by Mike-Leo-Smith** [reply]
BTW, do you mean initializing the future Vulkan backend, or the existing `VulkanSwapchain` things in the CUDA/CPU backends?

**Reply by trsh** [reply]
Anything that makes LuisaCompute work. I just dont want to create second Vulkan context, devices etc. My main APP is already powered by vulkan, and its API exposes existing `phy_device, device, vkGetInstanceProcAddr, vkGetDeviceProcAddr and vk_instance`. This is how it looks in another product I am thinkin on using:

```
ctx = new VKContext(); // helper
ctx ->create(vk_instance, (void*)vkGetInstanceProcAddr_, vk_physical_device, vk_device, vk_family, 0) // Init with existing context features
otherFrameworkDevice = Device(ctx ) // create other frame work device, and use it as always - this could be Luisa device in our case orsmth
```

**Reply by Mike-Leo-Smith** [reply, marked as answer]
Understood. We'll consider this when making the Vulkan backend.

**Reply by Mike-Leo-Smith** [reply]
Users on other backends might also have the same demands. We should design a unified way of importing existing devices.

**Comment by trsh** [comment]
> Users on other backends might also have the same demands. We should design a unified way of importing existing devices.

Yeah. In my case, I want to do some additional computation on GPU, and I don't want to go trough the hell of setting up 500 lines of vulkan code to run a compute kernel. Neither start new Vulkan context.
